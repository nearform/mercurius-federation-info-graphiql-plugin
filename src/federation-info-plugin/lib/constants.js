// introspectionTypes from GraphQL-JS repo: /src/type/introspection.ts
import { introspectionTypes, specifiedScalarTypes } from 'graphql'
export const INTROSPECTION_TYPES = introspectionTypes.map(({ name }) => name)

// specifiedScalarTypes from GraphQL-JS repo: /src/type/scalars.ts
export const SCALAR_TYPES = specifiedScalarTypes.map(({ name }) => name)

// Observed from examples
export const IGNORED_FIELDS = ['_entities', '_service']

// https://github.com/mercurius-js/mercurius/blob/57ff23ab97c8dec1b2f6fa0f6d0439b22cdb9377/lib/federation.js#L46 //_Any and _FieldSet
// https://github.com/mercurius-js/mercurius/blob/57ff23ab97c8dec1b2f6fa0f6d0439b22cdb9377/lib/federation.js#L174 //_Entity
// https://github.com/mercurius-js/mercurius/blob/57ff23ab97c8dec1b2f6fa0f6d0439b22cdb9377/lib/federation.js#L56 // _Service
export const BASE_FEDERATION_SCALARS = [
  '_Any',
  '_FieldSet',
  '_Entity',
  '_Service'
]

//rollup needs upgrading to support spread it seems
export const IGNORED_TYPES = [
  ...INTROSPECTION_TYPES,
  ...BASE_FEDERATION_SCALARS,
  ...SCALAR_TYPES
]
