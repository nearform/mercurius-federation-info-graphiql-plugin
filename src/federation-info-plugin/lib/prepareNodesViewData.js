import { IGNORED_FIELDS, IGNORED_TYPES } from './constants'
import introspectionTypeToString from './introspectionTypeToString'
import { TypeKind } from 'graphql'

/**
 * Reduces the list of fields into a { [name]: field } map, removes internal fields, attaches the nodeName and stringified type name to it
 *
 * @param {import('graphql').IntrospectionField[]} nodeFields
 * @param {string} nodeName
 * @returns {{[fieldName: string]: import('graphql').IntrospectionField & { nodeName: string, typeString: string }}}
 */
export const indexNodeFields = (nodeFields, nodeName) => {
  return nodeFields.reduce((result, field) => {
    if (!IGNORED_FIELDS.includes(field.name)) {
      field.nodeName = nodeName
      field.typeString = introspectionTypeToString(field.type)
      result[field.name] = field
    }
    return result
  }, {})
}

/**
 * Tranforms the object returned fromt the server into an array of objects that is easy to extract owner/reference information, without excessive iteration.
 * Ideally should replace `parseFederationSchema`
 *
 * @param {{[node: string]: { __schema: import('graphql').IntrospectionQuery } }} federationSchema example: { "node1": {__schema: IntrospectionQuery }}
 * @returns Object that is used to retrieve referencedBy and Owner
 * @example
 * ```
 * [
 * {
 *      nodeName:'node1',
 *      itemsMap: {
 *          User: {
 *              nodeName:'node1',
 *              itemsMap: {
 *                  typeString: '[Int]!',
 *                  id: {
 *                      nodeName:'node1'
 *                      },
 *                  ...rest of the field props
 *                  },
 *              ...rest of the type props
 *              },
 *          ...rest of the node props
 *      }
 * }
 * ]
 *```
 */
export const prepareNodesViewData = federationSchema => {
  const nodesEntries = Object.entries(federationSchema.nodes)
  return nodesEntries.map(([nodeName, { __schema }]) => {
    const { types } = __schema
    const itemsMap = types.reduce((result, type) => {
      if (!IGNORED_TYPES.includes(type.name)) {
        let itemsMap
        if (type.kind === TypeKind.OBJECT) {
          itemsMap = indexNodeFields(type.fields, nodeName)
        } else if (type.kind === TypeKind.INPUT_OBJECT) {
          itemsMap = indexNodeFields(type.inputFields, nodeName)
        } else {
          itemsMap = {}
        }

        result[type.name] = { ...type, itemsMap, nodeName }
      }
      return result
    }, {})

    return { nodeName, itemsMap }
  })
}
