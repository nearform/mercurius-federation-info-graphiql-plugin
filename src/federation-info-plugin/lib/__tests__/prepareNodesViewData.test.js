import { IGNORED_FIELDS } from '../constants'
import { buildSchema, introspectionFromSchema } from 'graphql'
import { indexNodeFields, prepareNodesViewData } from '../prepareNodesViewData'

const fieldsFromNames = names =>
  names.map(name => ({
    name,
    type: {
      name: 'String'
    }
  }))

describe('indexNodeFields', () => {
  it('returns indexed fields', () => {
    const fields = fieldsFromNames([...IGNORED_FIELDS, 'test'])

    expect(indexNodeFields(fields, 'dummyNode')).toEqual({
      test: {
        name: 'test',
        type: { name: 'String' },
        nodeName: 'dummyNode',
        typeString: 'String'
      }
    })
  })
})

describe('prepareNodesViewData', () => {
  it('Checks correct result', () => {
    const testNode = introspectionFromSchema(
      buildSchema(
        `
          type Query {
            test(customInput: CustomInput): CustomType
          }

          type CustomType {
            id: ID
            name: String
          }

          input CustomInput {
            names: [String]!
          }
          
          type _Entity {
            name: String
          }

          enum TestEnum {
            one
            two
            three
          }
        `
      )
    )
    const nodesResult = { nodes: { testNode } }
    expect(prepareNodesViewData(nodesResult)).toMatchSnapshot()

    // Potential replacments for snapshot test
    // Object.keys(itemsMap) equals ['Query', 'CustomType', 'CustomInput', 'TestEnum']

    // itemsMap.Query.itemsMap.test

    // itemsMap.CustomInput.itemsMap.names
    // itemsMap.CustomInput.itemsMap.names.typeString": "[String]!"
  })
})
