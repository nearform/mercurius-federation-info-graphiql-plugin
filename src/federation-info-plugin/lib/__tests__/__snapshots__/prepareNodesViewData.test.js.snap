// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareNodesViewData Checks correct result 1`] = `
Array [
  Object {
    "itemsMap": Object {
      "CustomInput": Object {
        "description": null,
        "enumValues": null,
        "fields": null,
        "inputFields": Array [
          Object {
            "defaultValue": null,
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "names",
            "nodeName": "testNode",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
            "typeString": "[String]!",
          },
        ],
        "interfaces": null,
        "itemsMap": Object {
          "names": Object {
            "defaultValue": null,
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "names",
            "nodeName": "testNode",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
            "typeString": "[String]!",
          },
        },
        "kind": "INPUT_OBJECT",
        "name": "CustomInput",
        "nodeName": "testNode",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
      "CustomType": Object {
        "description": null,
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "nodeName": "testNode",
            "type": Object {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null,
            },
            "typeString": "ID",
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "nodeName": "testNode",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
            "typeString": "String",
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "itemsMap": Object {
          "id": Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "nodeName": "testNode",
            "type": Object {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null,
            },
            "typeString": "ID",
          },
          "name": Object {
            "args": Array [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "nodeName": "testNode",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
            "typeString": "String",
          },
        },
        "kind": "OBJECT",
        "name": "CustomType",
        "nodeName": "testNode",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
      "Query": Object {
        "description": null,
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [
              Object {
                "defaultValue": null,
                "deprecationReason": null,
                "description": null,
                "isDeprecated": false,
                "name": "customInput",
                "type": Object {
                  "kind": "INPUT_OBJECT",
                  "name": "CustomInput",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "test",
            "nodeName": "testNode",
            "type": Object {
              "kind": "OBJECT",
              "name": "CustomType",
              "ofType": null,
            },
            "typeString": "CustomType",
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "itemsMap": Object {
          "test": Object {
            "args": Array [
              Object {
                "defaultValue": null,
                "deprecationReason": null,
                "description": null,
                "isDeprecated": false,
                "name": "customInput",
                "type": Object {
                  "kind": "INPUT_OBJECT",
                  "name": "CustomInput",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "test",
            "nodeName": "testNode",
            "type": Object {
              "kind": "OBJECT",
              "name": "CustomType",
              "ofType": null,
            },
            "typeString": "CustomType",
          },
        },
        "kind": "OBJECT",
        "name": "Query",
        "nodeName": "testNode",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
      "TestEnum": Object {
        "description": null,
        "enumValues": Array [
          Object {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "one",
          },
          Object {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "two",
          },
          Object {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "three",
          },
        ],
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "itemsMap": Object {},
        "kind": "ENUM",
        "name": "TestEnum",
        "nodeName": "testNode",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
    },
    "nodeName": "testNode",
  },
]
`;
